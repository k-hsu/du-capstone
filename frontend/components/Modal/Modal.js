import React, { useLayoutEffect, useState } from "react";
import { createPortal } from "react-dom";
import styled from "@emotion/styled";
import Button from "../Button/Button";
import Flex from "../Flex/Flex";
import Text from "../Text/Text";
import { color, spacing } from "../../theme";

const StyledModalBackground = styled.div`
  position: fixed;
  inset: 0;
  background-color: ${color.black60};
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledModal = styled.div`
  padding: ${spacing["2"]} ${spacing["2.25"]};
  background-color: ${color.white};
  min-width: 400px;
  border-radius: ${spacing["0.75"]};
`;

const Modal = ({
  title,
  children,
  onClose,
  onSubmit,
  cancelText = "Cancel",
  submitText = "Submit",
}) => {
  const [appRoot, setAppRoot] = useState(null);
  useLayoutEffect(() => {
    if (appRoot !== document?.getElementById("app-root")) {
      setAppRoot(document.getElementById("app-root"));
    }
  }, [appRoot]);
  return appRoot
    ? createPortal(
        <StyledModalBackground onClick={onClose}>
          <StyledModal onClick={(event) => event.stopPropagation()}>
            <Flex justifyContent="space-between" p={`0 0 ${spacing["2"]} 0`}>
              <Text fontWeight="bold">{title}</Text>
              <Text
                onClick={onClose}
                onKeyDown={onClose}
                cursor="pointer"
                tabIndex={0}
              >
                X
              </Text>
            </Flex>
            {children}
            <Flex
              justifyContent="flex-end"
              alignItems="center"
              gap={spacing["1.5"]}
              p={`${spacing["2"]} 0 0 0`}
            >
              <Text
                color={color.red}
                onClick={onClose}
                onKeyDown={onClose}
                lineHeight={0}
                cursor="pointer"
                tabIndex={0}
              >
                {cancelText}
              </Text>
              <Button onClick={onSubmit}>
                <Text>{submitText}</Text>
              </Button>
            </Flex>
          </StyledModal>
        </StyledModalBackground>,
        appRoot
      )
    : null;
};

export default Modal;
