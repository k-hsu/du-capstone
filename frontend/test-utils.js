import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderHook } from "@testing-library/react-hooks";

function setup(jsx) {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  };
}

export * from "@testing-library/react";
export { renderHook, setup };
