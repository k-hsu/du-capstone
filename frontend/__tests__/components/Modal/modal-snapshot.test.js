import React from "react";
import { renderer } from "../../../test-utils";
import Modal from "../../../components/Modal/Modal";

jest.mock("react-dom", () => {
  const originalModule = jest.requireActual("react-dom");

  return {
    ...originalModule,
    createPortal: (element) => element,
  };
});

describe("Modal Snapshot", () => {
  const spy = jest.spyOn(document, "getElementById");
  spy.mockReturnValue(<div id="app-root" />);
  it("should render modal component styles", () => {
    const component = renderer.create(
      <Modal title="Notification">
        You have been notified of a friend request
      </Modal>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
