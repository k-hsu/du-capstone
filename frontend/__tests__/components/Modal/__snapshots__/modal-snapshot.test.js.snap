// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Snapshot should render modal component styles 1`] = `
<div
  className="css-2wtgp9"
>
  <div
    className="css-15e9n3k"
    onClick={[Function]}
  >
    <div
      className="css-1pgkpxz"
      direction="row"
    >
      <p
        className="css-kxn7ez"
        fontFamily="IBM Plex Sans"
        fontWeight="bold"
      >
        Notification
      </p>
      <p
        className="css-7ee2u0"
        cursor="pointer"
        fontFamily="IBM Plex Sans"
        tabIndex={0}
      >
        X
      </p>
    </div>
    You have been notified of a friend request
    <div
      className="css-1a6mal2"
      direction="row"
    >
      <p
        className="css-13kh8s7"
        color="#F00"
        cursor="pointer"
        fontFamily="IBM Plex Sans"
        tabIndex={0}
      >
        Cancel
      </p>
      <button
        className="css-ixr4om"
      >
        <p
          className="css-ms6223"
          fontFamily="IBM Plex Sans"
        >
          Submit
        </p>
      </button>
    </div>
  </div>
</div>
`;
